In [3]:

#importamos librerías necesarias
# Python ≥3.5 is required
import sys
assert sys.version_info >= (3, 5)

# Scikit-Learn ≥0.20 is required
import sklearn
assert sklearn.__version__ >= "0.20"

# Common imports
import numpy as np
import os
import pandas as pd


# to make this notebook's output stable across runs
#np.random.seed(42)

# To plot pretty figures
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=20)
mpl.rc('xtick', labelsize=20)
mpl.rc('ytick', labelsize=20)

In [2]:
'''   
En este archivo vamos a unir los datos estadísticos extraídos.
Posteriormente, trataremos de asignar el valor de mercado correspondiente de cada jugador.
'''
#DATAFRAMES GENERALES
df_general_ESP = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/España2018.xls')
df_general_ING = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Inglaterra2018.xls')
df_general_ITA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Italia2018.xls')
df_general_ALE = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Alemania2018.xls')
df_general_FRA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Francia2018.xls')
df_general_POR = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Portugal2018.xls')
df_general_HOL = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_generales/Holanda2018.xls')

#DATAFRAMES PORTEROS
df_porteros_ESP = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/España2018_GK.xls')
df_porteros_ING = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Inglaterra2018_GK.xls')
df_porteros_ITA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Italia2018_GK.xls')
df_porteros_ALE = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Alemania2018_GK.xls')
df_porteros_FRA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Francia2018_GK.xls')
df_porteros_POR = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Portugal2018_GK.xls')
df_porteros_HOL = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_porteros/Holanda2018_GK.xls')

#DATAFRAMES GOLES ENCAJADOS
df_golesencajados_ESP = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/España2018GA.xlsx')
df_golesencajados_ING = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Inglaterra2018GA.xlsx')
df_golesencajados_ITA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Italia2018GA.xlsx')
df_golesencajados_ALE = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Alemania2018GA.xlsx')
df_golesencajados_FRA = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Francia2018GA.xlsx')
df_golesencajados_POR = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Portugal2018GA.xlsx')
df_golesencajados_HOL = pd.read_excel('/Users/alexcar934/Desktop/src/data/raw/estadísticas_golesencajados/Holanda2018GA.xlsx')


#DATAFRAME DE JUGADORES Y SU VALOR DE MERCADO EN EL VERANO DE 2019
df_market = pd.read_excel('/Users/alexcar934/Desktop/src/data/processed/MarketValues_Clean.xlsx')

df = df_general.dropna().merge(df_porteros.dropna(), how = 'outer', on = 'Player')
df = df.replace(np.nan, 0)
df = df.merge(df_defense.dropna(), how = 'outer', on = 'Player')
df = df.merge(df_pases.dropna(), how = 'outer', on = 'Player')
df = df.merge(df_regates.dropna(), how = 'outer', on = 'Player')
df = df.replace(np.nan, 0)
df.columns = ['Nombre del jugador','Posición','Equipo','Edad','Minutos jugados','Goles','Asistencias',
            'Goles encajados (Porteros)', 'Paradas (Porteros)', 'Porterías a cero (Porteros)', 'Bloqueos','Intercepciones','Despejes',
            'Pases acertados','Regates acertados']
df

# España
Vamos a crear un DataFrame con:

    1. Jugadores de La Liga 

    2. Estadísticas de los mismos en la temporada 2018-2019 

    3. Sus valores de mercado en el verano de 2019 

In [3]:
df_porteros_ESP.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_ESP.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']
df_españa = df_general_ESP.dropna().merge(df_porteros_ESP.dropna(),how= 'outer', on = 'Player')
df_españa = df_españa.replace(np.nan, 0)
df_españa = df_españa.merge(df_golesencajados_ESP.dropna(), how='outer', on='Player')


In [4]:
df_españa

Rk                 Player    Pos            Squad   Age  \
0    1.0           Antonio Adan     GK  Atlético Madrid  31.0   
1    2.0                   Aday  DF,MF           Girona  30.0   
2    3.0           Aritz Aduriz     FW    Athletic Club  37.0   
3    4.0         Luis Advincula     DF   Rayo Vallecano  28.0   
4    5.0  Martin Aguirregabiria     DF           Alavés  22.0   
..   ...                    ...    ...              ...   ...   
644  NaN            Ivan Villar    NaN              NaN   NaN   
645  NaN              Marc Vito    NaN              NaN   NaN   
646  NaN          Daniel Vivian    NaN              NaN   NaN   
647  NaN            Ruben Yanez    NaN              NaN   NaN   
648  NaN       Andoni Zubiaurre    NaN              NaN   NaN   

     Playing Time|Min  Performance|Gls  Performance|Ast   xG  \
0                90.0              0.0              0.0  0.0   
1               645.0              0.0              2.0  0.3   
2               943.0              2.0              0.0  3.3   
3              2389.0              1.0              3.0  1.1   
4              2032.0              0.0              0.0  0.0   
..                ...              ...              ...  ...   
644               NaN              NaN              NaN  NaN   
645               NaN              NaN              NaN  NaN   
646               NaN              NaN              NaN  NaN   
647               NaN              NaN              NaN  NaN   
648               NaN              NaN              NaN  NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           2.0                 5.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
644                         NaN                 NaN   
645                         NaN                 NaN   
646                         NaN                 NaN   
647                         NaN                 NaN   
648                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados  Goles encajados esperados  
0                            0.0                4                        2.7  
1                            0.0               10                        9.5  
2                            0.0               14                       13.2  
3                            0.0               48                       41.2  
4                            0.0               30                       31.6  
..                           ...              ...                        ...  
644                          NaN                0                        0.0  
645                          NaN                0                        0.0  
646                          NaN                0                        0.0  
647                          NaN                0                        0.0  
648                          NaN                0                        0.0  

[649 rows x 14 columns]

In [5]:
df_españa.columns = ['Rk','Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas', 'Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)',
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados' ]
df_españa = df_españa.dropna().iloc[:,1:]
df_españa

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0             Antonio Adan       GK  Atlético Madrid  31.0             90.0   
1                     Aday    DF,MF           Girona  30.0            645.0   
2             Aritz Aduriz       FW    Athletic Club  37.0            943.0   
3           Luis Advincula       DF   Rayo Vallecano  28.0           2389.0   
4    Martin Aguirregabiria       DF           Alavés  22.0           2032.0   
..                     ...      ...              ...   ...              ...   
579          Nabil El Zhar    MF,FW          Leganés  31.0            966.0   
580            Luca Zidane       GK      Real Madrid  20.0             90.0   
581          Igor Zubeldia       MF    Real Sociedad  21.0           2549.0   
582       Martin Zubimendi       MF    Real Sociedad  19.0              7.0   
583         David Zurutuza       MF    Real Sociedad  32.0           1222.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               0.0                   2.0                       0.3   
2               2.0                   0.0                       3.3   
3               1.0                   3.0                       1.1   
4               0.0                   0.0                       0.0   
..              ...                   ...                       ...   
579             4.0                   0.0                       2.9   
580             0.0                   0.0                       0.0   
581             0.0                   1.0                       0.5   
582             0.0                   0.0                       0.0   
583             2.0                   0.0                       1.6   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           2.0                 5.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
579                         0.0                 0.0   
580                         2.0                 1.0   
581                         0.0                 0.0   
582                         0.0                 0.0   
583                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                        4   
1                            0.0                       10   
2                            0.0                       14   
3                            0.0                       48   
4                            0.0                       30   
..                           ...                      ...   
579                          0.0                       15   
580                          0.0                        2   
581                          0.0                       37   
582                          0.0                        1   
583                          0.0                       15   

     Goles encajados esperados  
0                          2.7  
1                          9.5  
2                         13.2  
3                         41.2  
4                         31.6  
..                         ...  
579                       12.7  
580                        1.7  
581                       36.8  
582                        0.0  
583                       16.7  

[584 rows x 13 columns]

In [6]:
df_market_españa = df_market[df_market['Liga']== 'LaLiga']
df_market_españa = df_market_españa.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_españa.drop_duplicates()

Nombre del jugador  Valor de Mercado
141        Adrian Marin            800000
142         Aleix Vidal           4000000
143      Antonio Sivera           1500000
144              Burgui           1500000
145       Daniel Torres            900000
...                 ...               ...
3314      Santi Cazorla           4000000
3315   Santiago Caseres          10000000
3316      Sergio Asenjo          12000000
3317     Vicente Iborra           9000000
3318     Xavi Quintilla           2000000

[489 rows x 2 columns]

In [7]:
df_españa= df_españa.merge(df_market_españa, on='Nombre del jugador')
df_españa = df_españa.drop_duplicates(subset='Nombre del jugador')

In [8]:
df_españa['Liga'] = 2
df_españa

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0             Antonio Adan       GK  Atlético Madrid  31.0             90.0   
1             Aritz Aduriz       FW    Athletic Club  37.0            943.0   
2    Martin Aguirregabiria       DF           Alavés  22.0           2032.0   
3               Jordi Alba       DF        Barcelona  29.0           2991.0   
4            Ruben Alcaraz       MF       Valladolid  27.0           2976.0   
..                     ...      ...              ...   ...              ...   
346         Inaki Williams       FW    Athletic Club  24.0           3025.0   
347           Okay Yokuslu       MF       Celta Vigo  24.0           2198.0   
348          Joseba Zaldua       DF    Real Sociedad  26.0           2139.0   
349          Igor Zubeldia       MF    Real Sociedad  21.0           2549.0   
350         David Zurutuza       MF    Real Sociedad  32.0           1222.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               2.0                   0.0                       3.3   
2               0.0                   0.0                       0.0   
3               2.0                   8.0                       1.8   
4               3.0                   2.0                       3.5   
..              ...                   ...                       ...   
346            13.0                   4.0                      13.1   
347             2.0                   0.0                       1.7   
348             1.0                   2.0                       0.1   
349             0.0                   1.0                       0.5   
350             2.0                   0.0                       1.6   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           2.0                 5.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
346                         0.0                 0.0   
347                         0.0                 0.0   
348                         0.0                 0.0   
349                         0.0                 0.0   
350                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                        4   
1                            0.0                       14   
2                            0.0                       30   
3                            0.0                       30   
4                            0.0                       43   
..                           ...                      ...   
346                          0.0                       39   
347                          0.0                       45   
348                          0.0                       29   
349                          0.0                       37   
350                          0.0                       15   

     Goles encajados esperados  Valor de Mercado  Liga  
0                          2.7           1500000     2  
1                         13.2           1000000     2  
2                         31.6           7500000     2  
3                         38.6          50000000     2  
4                         49.6          10000000     2  
..                         ...               ...   ...  
346                       38.7          50000000     2  
347                       41.4          12000000     2  
348                       31.0          10000000     2  
349                       36.8          30000000     2  
350                       16.7           2500000     2  

[298 rows x 15 columns]

# Inglaterra

In [9]:
df_porteros_ING.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_ING.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']
df_inglaterra = df_general_ING.dropna().merge(df_porteros_ING.dropna(),how= 'outer', on = 'Player')
df_inglaterra = df_inglaterra.replace(np.nan, 0)
df_inglaterra  = df_inglaterra.merge(df_golesencajados_ING.dropna(), how='outer', on='Player')

In [10]:
df_inglaterra.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas', 'Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)',
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados' ]
df_inglaterra

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0    Patrick Van Aanholt       DF   Crystal Palace  27.0           3195.0   
1          Sergio Aguero       FW  Manchester City  30.0           2459.0   
2             Nathan Ake       DF      Bournemouth  23.0           3411.0   
3        Marc Albrighton    FW,MF   Leicester City  28.0           1625.0   
4      Toby Alderweireld       DF        Tottenham  29.0           2922.0   
..                   ...      ...              ...   ...              ...   
576            Ben White      NaN              NaN   NaN              NaN   
577       Alfie Whiteman      NaN              NaN   NaN              NaN   
578      Freddie Woodman      NaN              NaN   NaN              NaN   
579            Sam Woods      NaN              NaN   NaN              NaN   
580         Maya Yoshida      NaN              NaN   NaN              NaN   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               3.0                   2.0                       2.6   
1              21.0                   8.0                      18.1   
2               4.0                   0.0                       3.4   
3               2.0                   2.0                       1.2   
4               0.0                   0.0                       1.1   
..              ...                   ...                       ...   
576             NaN                   NaN                       NaN   
577             NaN                   NaN                       NaN   
578             NaN                   NaN                       NaN   
579             NaN                   NaN                       NaN   
580             NaN                   NaN                       NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
576                         NaN                 NaN   
577                         NaN                 NaN   
578                         NaN                 NaN   
579                         NaN                 NaN   
580                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                       46   
1                            0.0                       15   
2                            0.0                       70   
3                            0.0                       17   
4                            0.0                       31   
..                           ...                      ...   
576                          NaN                        0   
577                          NaN                        0   
578                          NaN                        0   
579                          NaN                        0   
580                          NaN                       29   

     Goles encajados esperados  
0                         46.3  
1                         17.7  
2                         57.1  
3                         23.1  
4                         37.0  
..                         ...  
576                        0.0  
577                        0.0  
578                        0.0  
579                        0.0  
580                       22.0  

[581 rows x 13 columns]

In [11]:
df_market_inglaterra = df_market[df_market['Liga']== 'Premier League']
df_market_inglaterra = df_market_inglaterra.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_inglaterra.drop_duplicates()

Nombre del jugador  Valor de Mercado
165   Ainsley Maitland-Niles          18000000
166      Alexandre Lacazette          70000000
167               Bernd Leno          30000000
168              Bukayo Saka           7000000
169           Calum Chambers          14000000
...                      ...               ...
3565             Ruben Neves          50000000
3566           Ruben Vinagre          10000000
3567            Rui Patricio          20000000
3568            Ryan Bennett           7000000
3569              Willy Boly          15000000

[503 rows x 2 columns]

In [12]:
df_inglaterra= df_inglaterra.merge(df_market_inglaterra, on='Nombre del jugador')
df_inglaterra = df_inglaterra.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)

In [13]:
df_inglaterra['Liga'] = 1
df_inglaterra

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0    Patrick Van Aanholt       DF   Crystal Palace  27.0           3195.0   
1          Sergio Aguero       FW  Manchester City  30.0           2459.0   
2             Nathan Ake       DF      Bournemouth  23.0           3411.0   
3        Marc Albrighton    FW,MF   Leicester City  28.0           1625.0   
4      Toby Alderweireld       DF        Tottenham  29.0           2922.0   
..                   ...      ...              ...   ...              ...   
313         Ashley Young       DF   Manchester Utd  33.0           2569.0   
314       Pablo Zabaleta       DF         West Ham  33.0           2092.0   
315        Wilfried Zaha       FW   Crystal Palace  25.0           3036.0   
316  Oleksandr Zinchenko       DF  Manchester City  21.0           1151.0   
317           Kurt Zouma       DF          Everton  23.0           2612.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               3.0                   2.0                       2.6   
1              21.0                   8.0                      18.1   
2               4.0                   0.0                       3.4   
3               2.0                   2.0                       1.2   
4               0.0                   0.0                       1.1   
..              ...                   ...                       ...   
313             2.0                   2.0                       1.9   
314             0.0                   1.0                       0.5   
315            10.0                   5.0                       0.2   
316             0.0                   3.0                       0.1   
317             2.0                   2.0                       1.7   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
313                         0.0                 0.0   
314                         0.0                 0.0   
315                         0.0                 0.0   
316                         0.0                 0.0   
317                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                       46   
1                            0.0                       15   
2                            0.0                       70   
3                            0.0                       17   
4                            0.0                       31   
..                           ...                      ...   
313                          0.0                       43   
314                          0.0                       35   
315                          0.0                       49   
316                          0.0                        4   
317                          0.0                       39   

     Goles encajados esperados  Valor de Mercado  Liga  
0                         46.3          10000000     1  
1                         17.7          65000000     1  
2                         57.1          35000000     1  
3                         23.1          10000000     1  
4                         37.0          40000000     1  
..                         ...               ...   ...  
313                       38.2           5000000     1  
314                       33.4           2000000     1  
315                       43.1          55000000     1  
316                        5.4          30000000     1  
317                       33.3          25000000     1  

[318 rows x 15 columns]

# Italia

In [14]:
df_porteros_ITA.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_ITA.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']
df_italia = df_general_ITA.dropna().merge(df_porteros_ITA.dropna(),how= 'outer', on = 'Player')
df_italia= df_italia.replace(np.nan, 0)
df_italia  = df_italia.merge(df_golesencajados_ITA.dropna(), how='outer', on='Player')

In [15]:
df_italia.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas', 'Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)',
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados' ]
df_italia

Nombre del jugador Posición    Equipo  Edad  Minutos jugados  \
0        Ignazio Abate       DF     Milan  31.0           1302.0   
1     Francesco Acerbi       DF     Lazio  30.0           3309.0   
2       Afriyie Acquah       CC    Empoli  26.0            151.0   
3       Claud Adjapong    CC,DL  Sassuolo  20.0            270.0   
4     Ali Adnan Kadhim       DF  Atalanta  24.0            166.0   
..                 ...      ...       ...   ...              ...   
771      Mattia Vitale      NaN       NaN   NaN              NaN   
772        Rok Vodisek      NaN       NaN   NaN              NaN   
773    Jamie Yayi Mpie      NaN       NaN   NaN              NaN   
774      Karim Zedadka      NaN       NaN   NaN              NaN   
775      Mauro Zitelli      NaN       NaN   NaN              NaN   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.1   
1               3.0                   1.0                       3.4   
2               2.0                   0.0                       1.4   
3               1.0                   0.0                       0.7   
4               0.0                   0.0                       0.2   
..              ...                   ...                       ...   
771             NaN                   NaN                       NaN   
772             NaN                   NaN                       NaN   
773             NaN                   NaN                       NaN   
774             NaN                   NaN                       NaN   
775             NaN                   NaN                       NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
771                         NaN                 NaN   
772                         NaN                 NaN   
773                         NaN                 NaN   
774                         NaN                 NaN   
775                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                     13.0   
1                            0.0                     44.0   
2                            0.0                     29.0   
3                            0.0                      2.0   
4                            0.0                      5.0   
..                           ...                      ...   
771                          NaN                      0.0   
772                          NaN                      0.0   
773                          NaN                      0.0   
774                          NaN                      0.0   
775                          NaN                      0.0   

     Goles encajados esperados  
0                         14.2  
1                         45.2  
2                         25.2  
3                          4.1  
4                          2.9  
..                         ...  
771                        0.0  
772                        0.0  
773                        0.0  
774                        0.0  
775                        0.0  

[776 rows x 13 columns]

In [16]:
df_market_italia = df_market[df_market['Liga']== 'Serie A']
df_market_italia = df_market_italia.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_italia.drop_duplicates()

Nombre del jugador  Valor de Mercado
60       Alessio Romagnoli          50000000
61            Andrea Conti          15000000
62              Ante Rebic          40000000
63      Antonio Donnarumma            500000
64         Davide Calabria          18000000
...                    ...               ...
3205     Sebastien De Maio           1000000
3206           Seko Fofana          13000000
3207         Stefano Okaka           5000000
3208                Walace           8500000
3209  William Troost-Ekong           4000000

[547 rows x 2 columns]

In [17]:
df_italia= df_italia.merge(df_market_italia, on='Nombre del jugador')
df_italia = df_italia.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)

In [18]:
df_italia['Liga'] = 3
df_italia

Nombre del jugador Posición    Equipo  Edad  Minutos jugados  \
0     Francesco Acerbi       DF     Lazio  30.0           3309.0   
1       Claud Adjapong    CC,DL  Sassuolo  20.0            270.0   
2             Ola Aina       DF    Torino  21.0            216.0   
3         Luis Alberto    CC,DL     Lazio  25.0            199.0   
4                Allan       CC    Napoli  27.0           2616.0   
..                 ...      ...       ...   ...              ...   
337     Nicolo Zaniolo    CC,DL      Roma  19.0           1766.0   
338    Cristian Zapata       DF     Milan  31.0           1026.0   
339       Duvan Zapata       DL  Atalanta  27.0           2896.0   
340        Simone Zaza    DL,CC    Torino  27.0           1437.0   
341    Piotr Zielinski       CC    Napoli  24.0           2875.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               3.0                   1.0                       3.4   
1               1.0                   0.0                       0.7   
2               1.0                   3.0                       1.4   
3               4.0                   5.0                       4.2   
4               1.0                   3.0                       1.0   
..              ...                   ...                       ...   
337             4.0                   2.0                       4.4   
338             0.0                   0.0                       0.2   
339            23.0                   7.0                      18.9   
340             4.0                   3.0                       6.6   
341             6.0                   2.0                       5.5   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
337                         0.0                 0.0   
338                         0.0                 0.0   
339                         0.0                 0.0   
340                         0.0                 0.0   
341                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                     44.0   
1                            0.0                      2.0   
2                            0.0                     23.0   
3                            0.0                     22.0   
4                            0.0                     29.0   
..                           ...                      ...   
337                          0.0                     23.0   
338                          0.0                      9.0   
339                          0.0                     40.0   
340                          0.0                     17.0   
341                          0.0                     30.0   

     Goles encajados esperados  Valor de Mercado  Liga  
0                         45.2          12000000     3  
1                          4.1           3500000     3  
2                         31.7          10000000     3  
3                         23.3          28000000     3  
4                         30.5          60000000     3  
..                         ...               ...   ...  
337                       27.6          40000000     3  
338                        9.6           2000000     3  
339                       35.6          45000000     3  
340                       21.5          10000000     3  
341                       29.4          40000000     3  

[342 rows x 15 columns]

# Alemania

In [19]:
df_porteros_ALE.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_ALE.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']
df_alemania = df_general_ALE.dropna().merge(df_porteros_ALE.dropna(),how= 'outer', on = 'Player')
df_alemania= df_alemania.replace(np.nan, 0)
df_alemania= df_alemania.merge(df_golesencajados_ALE.dropna(), how='outer', on='Player')

In [20]:
df_alemania.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas', 'Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)',
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados' ]
df_alemania

Nombre del jugador Posición          Equipo  Edad  Minutos jugados  \
0          Issah Abbas       FW        Mainz 05  19.0              9.0   
1        David Abraham       DF  Eint Frankfurt  32.0            142.0   
2         Amir Abrashi       MF        Freiburg  28.0            790.0   
3          Tyler Adams       MF      RB Leipzig  19.0            721.0   
4      Antonis Aidonis       DF       Stuttgart  17.0             80.0   
..                 ...      ...             ...   ...              ...   
545     Niklas Wiemann      NaN             NaN   NaN              NaN   
546          Paul Will      NaN             NaN   NaN              NaN   
547       Raphael Wolf      NaN             NaN   NaN              NaN   
548      Robin Ziegele      NaN             NaN   NaN              NaN   
549     Jan Zimmermann      NaN             NaN   NaN              NaN   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               0.0                   0.0                       0.8   
2               0.0                   0.0                       0.2   
3               0.0                   2.0                       0.1   
4               0.0                   0.0                       0.0   
..              ...                   ...                       ...   
545             NaN                   NaN                       NaN   
546             NaN                   NaN                       NaN   
547             NaN                   NaN                       NaN   
548             NaN                   NaN                       NaN   
549             NaN                   NaN                       NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
545                         NaN                 NaN   
546                         NaN                 NaN   
547                         NaN                 NaN   
548                         NaN                 NaN   
549                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                        0   
1                            0.0                       29   
2                            0.0                       15   
3                            0.0                        2   
4                            0.0                        2   
..                           ...                      ...   
545                          NaN                        0   
546                          NaN                        0   
547                          NaN                        0   
548                          NaN                        0   
549                          NaN                        0   

     Goles encajados esperados  
0                          0.1  
1                         26.6  
2                         16.2  
3                          6.9  
4                          1.3  
..                         ...  
545                        0.0  
546                        0.0  
547                        0.0  
548                        0.0  
549                        0.0  

[550 rows x 13 columns]

In [21]:
df_market_alemania = df_market[df_market['Liga']== 'Bundesliga']
df_market_alemania = df_market_alemania.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_alemania.drop_duplicates()

Nombre del jugador  Valor de Mercado
0            Anthony Modeste           7000000
1              Benno Schmitz           1000000
2          Birger Verstraete           4000000
3                Brady Scott            100000
4          Christian Clemens           1750000
...                      ...               ...
3489      Sebastian Langkamp           2000000
3490           Simon Straudi            200000
3491         Stefanos Kapino           1000000
3492  Theodor Gebre Selassie           2500000
3493              Yuya Osako           5000000

[531 rows x 2 columns]

In [22]:
df_alemania= df_alemania.merge(df_market_alemania, on='Nombre del jugador')
df_alemania = df_alemania.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)

In [23]:
df_alemania['Liga'] = 4
df_alemania

Nombre del jugador Posición          Equipo  Edad  Minutos jugados  \
0          David Abraham       DF  Eint Frankfurt  32.0            142.0   
1           Amir Abrashi       MF        Freiburg  28.0            790.0   
2            Tyler Adams       MF      RB Leipzig  19.0            721.0   
3          Manuel Akanji       DF        Dortmund  23.0           2205.0   
4         Kevin Akpoguma       DF      Hoffenheim  23.0            499.0   
..                   ...      ...             ...   ...              ...   
299        Denis Zakaria       MF      M'Gladbach  21.0           1959.0   
300        Robin Zentner       GK        Mainz 05  23.0            900.0   
301          Jean Zimmer    DF,FW      Düsseldorf  24.0           1435.0   
302  Matthias Zimmermann    DF,MF      Düsseldorf  26.0           3049.0   
303         Steven Zuber    MF,FW      Hoffenheim  26.0            571.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.8   
1               0.0                   0.0                       0.2   
2               0.0                   2.0                       0.1   
3               1.0                   0.0                       0.5   
4               0.0                   0.0                       0.0   
..              ...                   ...                       ...   
299             4.0                   0.0                       3.8   
300             0.0                   0.0                       0.0   
301             1.0                   2.0                       0.5   
302             1.0                   3.0                       0.5   
303             0.0                   1.0                       0.6   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
299                         0.0                 0.0   
300                        14.0                35.0   
301                         0.0                 0.0   
302                         0.0                 0.0   
303                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                       29   
1                            0.0                       15   
2                            0.0                        2   
3                            0.0                       35   
4                            0.0                       12   
..                           ...                      ...   
299                          0.0                       26   
300                          2.0                       14   
301                          0.0                       33   
302                          0.0                       65   
303                          0.0                        8   

     Goles encajados esperados  Valor de Mercado  Liga  
0                         26.6           2500000     4  
1                         16.2           2500000     4  
2                          6.9          13000000     4  
3                         30.3          40000000     4  
4                         12.3           4000000     4  
..                         ...               ...   ...  
299                       33.8          28000000     4  
300                       17.7           1000000     4  
301                       30.0           2750000     4  
302                       62.3           5000000     4  
303                        8.9           7000000     4  

[304 rows x 15 columns]

# Francia

In [24]:
df_porteros_FRA.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_FRA.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']
df_francia = df_general_FRA.dropna().merge(df_porteros_FRA.dropna(),how= 'outer', on = 'Player')
df_francia= df_francia.replace(np.nan, 0)
df_francia= df_francia.merge(df_golesencajados_FRA.dropna(), how='outer', on='Player')

In [25]:
df_francia.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas', 'Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)',
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados' ]
df_francia

Nombre del jugador Posición         Equipo  Edad  Minutos jugados  \
0             Ismael Aaneba       DF     Strasbourg  19.0            204.0   
1          Yunis Abdelhamid       DF          Reims  30.0            342.0   
2               Mehdi Abeid       CC          Dijon  25.0           2023.0   
3               Charles Abi    DF,DL  Saint-Étienne  18.0             32.0   
4             Khaled Adenon       DF         Amiens  33.0           1994.0   
..                      ...      ...            ...   ...              ...   
676        Baptiste Valette      NaN            NaN   NaN              NaN   
677              Marc Vidal      NaN            NaN   NaN              NaN   
678  Panagiotis Vlachodimos      NaN            NaN   NaN              NaN   
679            Arthur Zagre      NaN            NaN   NaN              NaN   
680           Erwin Zelazny      NaN            NaN   NaN              NaN   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               0.0                   0.0                       1.6   
2               3.0                   0.0                       2.9   
3               0.0                   0.0                       0.1   
4               0.0                   0.0                       0.7   
..              ...                   ...                       ...   
676             NaN                   NaN                       NaN   
677             NaN                   NaN                       NaN   
678             NaN                   NaN                       NaN   
679             NaN                   NaN                       NaN   
680             NaN                   NaN                       NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
676                         NaN                 NaN   
677                         NaN                 NaN   
678                         NaN                 NaN   
679                         NaN                 NaN   
680                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                      2.0   
1                            0.0                     42.0   
2                            0.0                     30.0   
3                            0.0                      0.0   
4                            0.0                     32.0   
..                           ...                      ...   
676                          NaN                      0.0   
677                          NaN                      0.0   
678                          NaN                      0.0   
679                          NaN                      0.0   
680                          NaN                      0.0   

     Goles encajados esperados  
0                          3.1  
1                         45.0  
2                         31.3  
3                          1.3  
4                         28.1  
..                         ...  
676                        0.0  
677                        0.0  
678                        0.0  
679                        0.0  
680                        0.0  

[681 rows x 13 columns]

In [26]:
df_market_francia = df_market[df_market['Liga']== 'Ligue 1']
df_market_francia = df_market_francia.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_francia.drop_duplicates()

Nombre del jugador  Valor de Mercado
860         Alfred Gomis           2000000
861   Bruno Ecuele Manga           3000000
862        Bryan Soumare            400000
863         Didier Ndong           4000000
864        Enzo Loiodice           5500000
...                  ...               ...
3152       Stephane Zobo            150000
3153      Steven Moreira           2000000
3154         Wesley Said           7500000
3155   William Vainqueur           1250000
3156         Yaya Sanogo           2500000

[539 rows x 2 columns]

In [27]:
df_francia= df_francia.merge(df_market_francia, on='Nombre del jugador')
df_francia = df_francia.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)

In [28]:
df_francia['Liga'] = 5
df_francia

Nombre del jugador Posición         Equipo  Edad  Minutos jugados  \
0        Ismael Aaneba       DF     Strasbourg  19.0            204.0   
1     Yunis Abdelhamid       DF          Reims  30.0            342.0   
2          Mehdi Abeid       CC          Dijon  25.0           2023.0   
3          Charles Abi    DF,DL  Saint-Étienne  18.0             32.0   
4          Yacine Adli       CC       Bordeaux  18.0            216.0   
..                 ...      ...            ...   ...              ...   
315               Xeka       CC          Lille  23.0            226.0   
316    Zaydou Youssouf    CC,DL       Bordeaux  19.0            422.0   
317       Arber Zeneli       DL          Reims  23.0            810.0   
318         Kevin Zohi    DL,CC     Strasbourg  21.0            627.0   
319      Bongani Zungu       CC         Amiens  25.0            230.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               0.0                   0.0                       1.6   
2               3.0                   0.0                       2.9   
3               0.0                   0.0                       0.1   
4               0.0                   0.0                       0.0   
..              ...                   ...                       ...   
315             2.0                   0.0                       1.7   
316             0.0                   0.0                       0.3   
317             3.0                   3.0                       3.1   
318             1.0                   1.0                       1.0   
319             0.0                   0.0                       0.1   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
315                         0.0                 0.0   
316                         0.0                 0.0   
317                         0.0                 0.0   
318                         0.0                 0.0   
319                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                      2.0   
1                            0.0                     42.0   
2                            0.0                     30.0   
3                            0.0                      0.0   
4                            0.0                      2.0   
..                           ...                      ...   
315                          0.0                     19.0   
316                          0.0                      6.0   
317                          0.0                     15.0   
318                          0.0                      4.0   
319                          0.0                      3.0   

     Goles encajados esperados  Valor de Mercado  Liga  
0                          3.1           1000000     5  
1                         45.0           2500000     5  
2                         31.3           1750000     5  
3                          1.3            500000     5  
4                          2.9           5500000     5  
..                         ...               ...   ...  
315                       26.8          10000000     5  
316                        6.4           4000000     5  
317                       13.3           6000000     5  
318                        8.7           2000000     5  
319                        2.7           2500000     5  

[320 rows x 15 columns]

# Portugal

In [29]:
df_porteros_POR.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_POR.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']

df_portugal = df_general_POR.dropna().merge(df_porteros_POR.dropna(),how='outer', on = 'Player')
df_portugal = df_portugal.replace(np.nan, 0)
df_portugal = df_portugal.merge(df_golesencajados_POR.dropna(), how = 'outer',on='Player')

In [30]:
df_portugal.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas','Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)', 
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados']
df_portugal

Nombre del jugador Posición       Equipo  Edad  Minutos jugados  \
0     Mohamed Abarhoun       DF   Moreirense  29.0           1171.0   
1       Amir Abedzadeh       GK     Marítimo  25.0            117.0   
2    Vincent Aboubakar       FW        Porto  26.0            456.0   
3      Danildo Accioly       DF  Santa Clara  37.0            909.0   
4            Yaw Ackah       MF     Boavista  19.0             18.0   
..                 ...      ...          ...   ...              ...   
614     Mamadou Traore      NaN          NaN   NaN              NaN   
615        Joao Valido      NaN          NaN   NaN              NaN   
616     Adelcio Varela      NaN          NaN   NaN              NaN   
617       Bruno Varela      NaN          NaN   NaN              NaN   
618     Robinho Vieira      NaN          NaN   NaN              NaN   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.1   
1               0.0                   0.0                       0.0   
2               4.0                   1.0                       4.1   
3               0.0                   0.0                       1.0   
4               0.0                   0.0                       0.0   
..              ...                   ...                       ...   
614             NaN                   NaN                       NaN   
615             NaN                   NaN                       NaN   
616             NaN                   NaN                       NaN   
617             NaN                   NaN                       NaN   
618             NaN                   NaN                       NaN   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           0.0                 0.0   
1                          16.0                38.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
614                         NaN                 NaN   
615                         NaN                 NaN   
616                         NaN                 NaN   
617                         NaN                 NaN   
618                         NaN                 NaN   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                     15.0   
1                            3.0                     16.0   
2                            0.0                      2.0   
3                            0.0                     18.0   
4                            0.0                      0.0   
..                           ...                      ...   
614                          NaN                      0.0   
615                          NaN                      0.0   
616                          NaN                      0.0   
617                          NaN                      0.0   
618                          NaN                      0.0   

     Goles encajados esperados  
0                         20.7  
1                         19.0  
2                          2.7  
3                         15.4  
4                          0.1  
..                         ...  
614                        0.0  
615                        0.0  
616                        0.0  
617                        0.0  
618                        0.0  

[619 rows x 13 columns]

In [31]:
df_market_portugal = df_market[df_market['Liga']== 'Liga NOS']
df_market_portugal = df_market_portugal.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_portugal.drop_duplicates()

Nombre del jugador  Valor de Mercado
384       Andre Moreira            800000
385        Andre Santos            800000
386         Andre Sousa            700000
387               Benny            850000
388          Chima Akas            300000
...                 ...               ...
3402         Nuno Pinto            200000
3403       Nuno Valente            500000
3404             Silvio            250000
3405       Tiago Castro            500000
3406           Zequinha            200000

[477 rows x 2 columns]

In [32]:
df_portugal= df_portugal.merge(df_market_portugal, on='Nombre del jugador')
df_portugal = df_portugal.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)


In [33]:
df_portugal['Liga'] = 6
df_portugal

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0       Amir Abedzadeh       GK         Marítimo  25.0            117.0   
1    Vincent Aboubakar       FW            Porto  26.0            456.0   
2            Yaw Ackah       MF         Boavista  19.0             18.0   
3         Marcos Acuna    DF,MF      Sporting CP  26.0           2574.0   
4          Joao Afonso       DF          Vitória  28.0            450.0   
..                 ...      ...              ...   ...              ...   
197            Welthon       FW          Vitória  26.0            598.0   
198             Wendel       MF      Sporting CP  20.0           1638.0   
199              Xadas    MF,FW            Braga  20.0            207.0   
200           Zequinha    MF,FW  Vitória Setúbal  31.0            966.0   
201   Andrija Zivkovic    FW,MF          Benfica  22.0            667.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               4.0                   1.0                       4.1   
2               0.0                   0.0                       0.0   
3               1.0                   9.0                       1.3   
4               0.0                   1.0                       0.0   
..              ...                   ...                       ...   
197             1.0                   1.0                       1.9   
198             1.0                   1.0                       1.2   
199             1.0                   1.0                       0.5   
200             1.0                   1.0                       1.9   
201             0.0                   3.0                       0.5   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                          16.0                38.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
197                         0.0                 0.0   
198                         0.0                 0.0   
199                         0.0                 0.0   
200                         0.0                 0.0   
201                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            3.0                     16.0   
1                            0.0                      2.0   
2                            0.0                      0.0   
3                            0.0                     26.0   
4                            0.0                      9.0   
..                           ...                      ...   
197                          0.0                      5.0   
198                          0.0                     16.0   
199                          0.0                      0.0   
200                          0.0                      8.0   
201                          0.0                      6.0   

     Goles encajados esperados  Valor de Mercado  Liga  
0                         19.0           1000000     6  
1                          2.7          10000000     6  
2                          0.1            300000     6  
3                         28.1          18000000     6  
4                          4.9            500000     6  
..                         ...               ...   ...  
197                        7.5           1000000     6  
198                       21.6          10000000     6  
199                        1.8           4000000     6  
200                       11.0            200000     6  
201                        6.3           8000000     6  

[202 rows x 15 columns]

# Holanda

In [34]:
df_porteros_HOL.columns = ['Player', 'Goles encajados (Porteros)','Paradas (Porteros)', 'Porterías a cero (Porteros)']
df_golesencajados_HOL.columns = ['Player', 'Goles encajados', 'Goles encajados esperados']

df_holanda = df_general_HOL.dropna().merge(df_porteros_HOL.dropna(),how='outer', on = 'Player')
df_holanda = df_holanda.replace(np.nan, 0)
df_holanda = df_holanda.merge(df_golesencajados_HOL.dropna(), how = 'outer',on='Player')

In [35]:
df_holanda.columns = ['Nombre del jugador', 'Posición', 'Equipo', 'Edad', 'Minutos jugados', 'Goles anotados',
                    'Asistencias anotadas','Goles anotados esperados', 'Goles encajados (Porteros)','Paradas (Porteros)', 
                    'Porterías a cero (Porteros)','Goles encajados jugando', 'Goles encajados esperados']

In [36]:
df_market_holanda = df_market[df_market['Liga']== 'Eredivisie']
df_market_holanda = df_market_holanda.loc[:,['Nombre del jugador', 'Valor de Mercado']]
df_market_holanda.drop_duplicates()

Nombre del jugador  Valor de Mercado
85            Aaron Meijers            850000
86       Aleksandar Bjelica            350000
87         Bilal Ould-Chikh            300000
88    Crysencio Summerville            350000
89             Danny Bakker            800000
...                     ...               ...
3541       Rick Zuijderwijk             75000
3542       Sebastian Holmen           1250000
3543    Timon Wellenreuther           1250000
3544      Vangelis Pavlidis           1750000
3545  Victor Van Den Bogert            100000

[483 rows x 2 columns]

In [37]:
df_holanda= df_holanda.merge(df_market_holanda, on='Nombre del jugador')
df_holanda = df_holanda.drop_duplicates(subset='Nombre del jugador').dropna().reset_index(drop=True)

In [38]:
df_holanda['Liga'] = 7
df_holanda

Nombre del jugador Posición           Equipo  Edad  \
0           Zakaria Aboukhlal       DL    PSV Eindhoven  18.0   
1                Amir Absalem       DF        Groningen  21.0   
2                 Jafar Arias       DL            Emmen  23.0   
3                Yassin Ayoub       CC        Feyenoord  24.0   
4    Jean-Christophe Bahebeck       DL          Utrecht  25.0   
..                        ...      ...              ...   ...   
223        Deyovaisio Zeefuik       DF        Groningen  20.0   
224                 Agim Zeka       CC  Fortuna Sittard  19.0   
225              Hakim Ziyech    DL,CC             Ajax  25.0   
226               Jeroen Zoet       PO    PSV Eindhoven  27.0   
227           Robert Zwinkels       PO     ADO Den Haag  35.0   

     Minutos jugados  Goles anotados  Asistencias anotadas  \
0               11.0             0.0                   0.0   
1              163.0             0.0                   1.0   
2             1973.0             4.0                   0.0   
3              230.0             2.0                   3.0   
4              688.0             3.0                   2.0   
..               ...             ...                   ...   
223           2654.0             1.0                   1.0   
224             15.0             0.0                   0.0   
225           2401.0            16.0                  13.0   
226            306.0             0.0                   0.0   
227           1969.0             0.0                   0.0   

     Goles anotados esperados  Goles encajados (Porteros)  Paradas (Porteros)  \
0                         0.0                         0.0                 0.0   
1                         0.0                         0.0                 0.0   
2                         5.1                         0.0                 0.0   
3                         0.4                         0.0                 0.0   
4                         2.2                         0.0                 0.0   
..                        ...                         ...                 ...   
223                       1.5                         0.0                 0.0   
224                       0.0                         0.0                 0.0   
225                      10.8                         0.0                 0.0   
226                       0.0                        26.0                58.0   
227                       0.0                        36.0                92.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                      0.0   
1                            0.0                      2.0   
2                            0.0                     45.0   
3                            0.0                      6.0   
4                            0.0                     10.0   
..                           ...                      ...   
223                          0.0                     34.0   
224                          0.0                      1.0   
225                          0.0                     24.0   
226                         16.0                     26.0   
227                          7.0                     36.0   

     Goles encajados esperados  Valor de Mercado  Liga  
0                          0.0           1500000     7  
1                          2.6            400000     7  
2                         37.4            500000     7  
3                          5.8           2500000     7  
4                          9.4           1500000     7  
..                         ...               ...   ...  
223                       37.5           1500000     7  
224                        0.9            150000     7  
225                       27.6          40000000     7  
226                       25.8          12000000     7  
227                       42.3            500000     7  

[228 rows x 15 columns]

# Unir los dataframes

In [39]:
df = pd.concat([df_españa,df_inglaterra,df_francia,df_italia,df_alemania,df_holanda,df_portugal])
df

Nombre del jugador Posición           Equipo  Edad  Minutos jugados  \
0             Antonio Adan       GK  Atlético Madrid  31.0             90.0   
1             Aritz Aduriz       FW    Athletic Club  37.0            943.0   
2    Martin Aguirregabiria       DF           Alavés  22.0           2032.0   
3               Jordi Alba       DF        Barcelona  29.0           2991.0   
4            Ruben Alcaraz       MF       Valladolid  27.0           2976.0   
..                     ...      ...              ...   ...              ...   
197                Welthon       FW          Vitória  26.0            598.0   
198                 Wendel       MF      Sporting CP  20.0           1638.0   
199                  Xadas    MF,FW            Braga  20.0            207.0   
200               Zequinha    MF,FW  Vitória Setúbal  31.0            966.0   
201       Andrija Zivkovic    FW,MF          Benfica  22.0            667.0   

     Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0               0.0                   0.0                       0.0   
1               2.0                   0.0                       3.3   
2               0.0                   0.0                       0.0   
3               2.0                   8.0                       1.8   
4               3.0                   2.0                       3.5   
..              ...                   ...                       ...   
197             1.0                   1.0                       1.9   
198             1.0                   1.0                       1.2   
199             1.0                   1.0                       0.5   
200             1.0                   1.0                       1.9   
201             0.0                   3.0                       0.5   

     Goles encajados (Porteros)  Paradas (Porteros)  \
0                           2.0                 5.0   
1                           0.0                 0.0   
2                           0.0                 0.0   
3                           0.0                 0.0   
4                           0.0                 0.0   
..                          ...                 ...   
197                         0.0                 0.0   
198                         0.0                 0.0   
199                         0.0                 0.0   
200                         0.0                 0.0   
201                         0.0                 0.0   

     Porterías a cero (Porteros)  Goles encajados jugando  \
0                            0.0                      4.0   
1                            0.0                     14.0   
2                            0.0                     30.0   
3                            0.0                     30.0   
4                            0.0                     43.0   
..                           ...                      ...   
197                          0.0                      5.0   
198                          0.0                     16.0   
199                          0.0                      0.0   
200                          0.0                      8.0   
201                          0.0                      6.0   

     Goles encajados esperados  Valor de Mercado  Liga  
0                          2.7           1500000     2  
1                         13.2           1000000     2  
2                         31.6           7500000     2  
3                         38.6          50000000     2  
4                         49.6          10000000     2  
..                         ...               ...   ...  
197                        7.5           1000000     6  
198                       21.6          10000000     6  
199                        1.8           4000000     6  
200                       11.0            200000     6  
201                        6.3           8000000     6  

[2012 rows x 15 columns]

# Enumerar los equipos según valor de mercado

### Problema: Los nombres de los equipos son distintos en un dataset y otro.

Solución: Usar la suma de valores de mercado obtenida en el DataFrame final como valor para crear el ranking

In [40]:
df_clubes = pd.DataFrame(df.groupby('Equipo')['Valor de Mercado'].sum().sort_values(ascending=False).reset_index())
df_clubes['Equipo (Ranking por Valor de Mercado)'] = range(1,135)
df_clubes = df_clubes.loc[:,['Equipo','Equipo (Ranking por Valor de Mercado)']]


In [41]:
'''  
Unimos el Ranking obtenido al DataFrame final
'''

df = df.merge(df_clubes, on='Equipo')

In [42]:
''' 
Sustituimos los nombres de los equipos por su posición en el ranking por valor de mercado.
Quitamos la columna de ranking por valor de mercado ya que estaría duplicada.
'''

df['Equipo'] = df['Equipo (Ranking por Valor de Mercado)']
df = df.iloc[:,:-1]
df

Nombre del jugador Posición  Equipo  Edad  Minutos jugados  \
0          Antonio Adan       GK       8  31.0             90.0   
1        Santiago Arias       DF       8  26.0           1772.0   
2          Angel Correa    MF,FW       8  23.0           1919.0   
3           Diego Costa       FW       8  29.0           1112.0   
4     Antoine Griezmann    FW,MF       8  27.0           3198.0   
...                 ...      ...     ...   ...              ...   
2007       Luther Singh    FW,MF     128  20.0            137.0   
2008       Luis Machado    FW,MF     129  25.0           1806.0   
2009      Andre Moreira       GK     129  22.0            450.0   
2010     Lawrence Ofori    MF,FW     129  20.0            288.0   
2011    Philipe Sampaio       DF     129  23.0            723.0   

      Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0                0.0                   0.0                       0.0   
1                1.0                   1.0                       0.4   
2                3.0                   2.0                       2.2   
3                2.0                   2.0                       1.7   
4               15.0                   8.0                      13.9   
...              ...                   ...                       ...   
2007             2.0                   0.0                       2.8   
2008             1.0                   0.0                       1.6   
2009             0.0                   0.0                       0.0   
2010             0.0                   0.0                       0.1   
2011             1.0                   0.0                       0.1   

      Goles encajados (Porteros)  Paradas (Porteros)  \
0                            2.0                 5.0   
1                            0.0                 0.0   
2                            0.0                 0.0   
3                            0.0                 0.0   
4                            0.0                 0.0   
...                          ...                 ...   
2007                         0.0                 0.0   
2008                         0.0                 0.0   
2009                        14.0                16.0   
2010                         0.0                 0.0   
2011                         0.0                 0.0   

      Porterías a cero (Porteros)  Goles encajados jugando  \
0                             0.0                      4.0   
1                             0.0                     10.0   
2                             0.0                     22.0   
3                             0.0                      8.0   
4                             0.0                     29.0   
...                           ...                      ...   
2007                          0.0                      0.0   
2008                          0.0                     32.0   
2009                          0.0                     14.0   
2010                          0.0                     11.0   
2011                          0.0                      8.0   

      Goles encajados esperados  Valor de Mercado  Liga  
0                           2.7           1500000     2  
1                          19.6          18000000     2  
2                          25.9          35000000     2  
3                          12.9          25000000     2  
4                          38.8         130000000     2  
...                         ...               ...   ...  
2007                        0.0           2000000     6  
2008                       33.3            500000     6  
2009                        7.4            800000     6  
2010                        5.5            300000     6  
2011                        8.3            900000     6  

[2012 rows x 15 columns]

# Problema serio

## Hay anomalías en los datos:

En la columna 'Minutos Jugados', hay ciertos jugadores de los que se ha extraído que han jugado 

(Por Ejemplo):

    Jan Oblak, portero titular del Atlético de Madrid, marca un registro de 333 minutos jugados cuando ha jugado 3330.
    Hay jugadores que registran 31 minutos cuando han jugado 3100.


## Solución propuesta:

Multiplicar por 100 los valores de la columna Minutos jugados: 

    Los jugadores que hayan jugado menos de 35 minutos
    SOLO si han encajado un valor igual o superior a 15 goles.

* En el caso de haber jugado 34 o 340 minutos, no has podido encajar 15 goles (en ese caso eres un outlier).
* Los valores que sean menores de 34 se multiplican por 100, por lo que pasan a haber jugado --> valores menores de 3400 minutos 
(tiempo suficiente para haber encajado 16 goles)

Multiplicar por 10 los valores de la columna Minutos jugados:

    Los jugadores que hayan jugado menos de 343 minutos
    SOLO si han encajado un valor igual o superior a 15 goles.

* En el caso de haber jugado 340 minutos, no has podido encajar 15 goles (en ese caso, eres un outlier)
* Los valores que sean menores de 343 se multiplican por 10, por lo que pasan a haber jugado --> valores menores de 3430 minutos 
(tiempo suficiente para haber encajado 16 goles).

* EXPLICACIÓN DE LOS VALORES:

Estas ligas las comprenden:
* Máximo 20 equipos
* Juegan entre ellos dos veces --> 38 partidos como máximo por equipo
* Los partidos son de 90 minutos --> un total de 3420 minutos por temporada.

* Que te metan 15 goles en 3 partidos y medio es complicado (342 minutos), en 3420 minutos... ¡No tanto!

In [43]:
minutos = []
for i in df.index:
    if np.all(df['Minutos jugados'][i] <= 34) and np.all(df['Goles encajados jugando'][i]>=15):
        minutos.append(df['Minutos jugados'][i]*100)
    elif np.all(df['Minutos jugados'][i]< 343) and np.all(df['Goles encajados jugando'][i]>=15):
        minutos.append(df['Minutos jugados'][i]*10)
    else:
        minutos.append(df['Minutos jugados'][i]) 

df['Minutos jugados'] = minutos
df.sort_values(by='Minutos jugados').head()

Nombre del jugador Posición  Equipo  Edad  Minutos jugados  \
995    Vincenzo Millico       CC      49  17.0              1.0   
979   Giacomo Raspadori       DF      38  18.0              1.0   
1136    Francesco Rossi       PO      28  27.0              1.0   
778    Theo Sainte-Luce       CC      80  19.0              1.0   
931           Sanasi Sy       CC      94  19.0              1.0   

      Goles anotados  Asistencias anotadas  Goles anotados esperados  \
995              0.0                   0.0                       0.0   
979              0.0                   0.0                       0.0   
1136             0.0                   0.0                       0.0   
778              0.0                   0.0                       0.0   
931              0.0                   0.0                       0.0   

      Goles encajados (Porteros)  Paradas (Porteros)  \
995                          0.0                 0.0   
979                          0.0                 0.0   
1136                         0.0                 0.0   
778                          0.0                 0.0   
931                          0.0                 0.0   

      Porterías a cero (Porteros)  Goles encajados jugando  \
995                           0.0                      0.0   
979                           0.0                      0.0   
1136                          0.0                      0.0   
778                           0.0                      0.0   
931                           0.0                      0.0   

      Goles encajados esperados  Valor de Mercado  Liga  
995                         0.0           1000000     3  
979                         0.0            500000     3  
1136                        0.0            100000     3  
778                         0.0            150000     5  
931                         0.0            150000     5

# La Posición

Son pocas, habrá que numerarlas según lo adelantada que esté dicha posición:

1. El portero
2. Defensas y laterales
3. Centrocampistas e interiores
4. Atacantes/Delanteros

In [44]:
df['Posición'].unique()

array(['GK', 'DF', 'MF,FW', 'FW', 'FW,MF', 'MF', 'MF,DF', 'DF,FW',
       'DF,MF', 'FW,DF', 'DL', 'DF,DL', 'CC,DL', 'CC', 'PO', 'CC,DF',
       'DL,CC', 'DF,CC', 'DL,DF'], dtype=object)

In [45]:
posicion=[]
for i in df['Posición']:
    if i == 'GK' or i == 'PO':
        posicion.append(1)
    elif i[0:2] == 'DF':
        posicion.append(2)
    elif i[0:2] == 'CC' or i[0:2] == 'MF':
        posicion.append(3)
    elif i[0:2] == 'DL' or i[0:2] == 'FW':
        posicion.append(4)
        
print(len(posicion))

2012


In [46]:
df['Posición'] = posicion
df

Nombre del jugador  Posición  Equipo  Edad  Minutos jugados  \
0          Antonio Adan         1       8  31.0             90.0   
1        Santiago Arias         2       8  26.0           1772.0   
2          Angel Correa         3       8  23.0           1919.0   
3           Diego Costa         4       8  29.0           1112.0   
4     Antoine Griezmann         4       8  27.0           3198.0   
...                 ...       ...     ...   ...              ...   
2007       Luther Singh         4     128  20.0            137.0   
2008       Luis Machado         4     129  25.0           1806.0   
2009      Andre Moreira         1     129  22.0            450.0   
2010     Lawrence Ofori         3     129  20.0            288.0   
2011    Philipe Sampaio         2     129  23.0            723.0   

      Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0                0.0                   0.0                       0.0   
1                1.0                   1.0                       0.4   
2                3.0                   2.0                       2.2   
3                2.0                   2.0                       1.7   
4               15.0                   8.0                      13.9   
...              ...                   ...                       ...   
2007             2.0                   0.0                       2.8   
2008             1.0                   0.0                       1.6   
2009             0.0                   0.0                       0.0   
2010             0.0                   0.0                       0.1   
2011             1.0                   0.0                       0.1   

      Goles encajados (Porteros)  Paradas (Porteros)  \
0                            2.0                 5.0   
1                            0.0                 0.0   
2                            0.0                 0.0   
3                            0.0                 0.0   
4                            0.0                 0.0   
...                          ...                 ...   
2007                         0.0                 0.0   
2008                         0.0                 0.0   
2009                        14.0                16.0   
2010                         0.0                 0.0   
2011                         0.0                 0.0   

      Porterías a cero (Porteros)  Goles encajados jugando  \
0                             0.0                      4.0   
1                             0.0                     10.0   
2                             0.0                     22.0   
3                             0.0                      8.0   
4                             0.0                     29.0   
...                           ...                      ...   
2007                          0.0                      0.0   
2008                          0.0                     32.0   
2009                          0.0                     14.0   
2010                          0.0                     11.0   
2011                          0.0                      8.0   

      Goles encajados esperados  Valor de Mercado  Liga  
0                           2.7           1500000     2  
1                          19.6          18000000     2  
2                          25.9          35000000     2  
3                          12.9          25000000     2  
4                          38.8         130000000     2  
...                         ...               ...   ...  
2007                        0.0           2000000     6  
2008                       33.3            500000     6  
2009                        7.4            800000     6  
2010                        5.5            300000     6  
2011                        8.3            900000     6  

[2012 rows x 15 columns]

### Por último...

# Variables que no interesan según a que posición

1. Los goles encajados por el equipo mientras el jugador está en el campo, así como los esperados, según lo adelantada que sea la posición del jugador, no interesa conocerlo.

2. Quitaremos los goles encajados por el equipo mientras el jugador está en el campo en el caso de los PORTEROS, por no repetir datos.

In [47]:
GA = []
for i in df.index:
    if np.all(df['Posición'][i]>=3) or np.all(df['Posición'][i] == 1):
        GA.append(0)
    else:
        GA.append(df['Goles encajados jugando'][i])

df['Goles encajados jugando'] = GA



In [48]:
GA_esp= []
for i in df.index:
    if np.all(df['Posición'][i]>=3) or np.all(df['Posición'][i] == 1):
        GA_esp.append(0)
    else:
        GA_esp.append(df['Goles encajados esperados'][i])

df['Goles encajados esperados'] = GA_esp
df.head(20)

Nombre del jugador  Posición  Equipo  Edad  Minutos jugados  \
0        Antonio Adan         1       8  31.0             90.0   
1      Santiago Arias         2       8  26.0           1772.0   
2        Angel Correa         3       8  23.0           1919.0   
3         Diego Costa         4       8  29.0           1112.0   
4   Antoine Griezmann         4       8  27.0           3198.0   
5                Koke         3       8  26.0           2586.0   
6        Thomas Lemar         3       8  22.0           1817.0   
7       Alvaro Morata         4       8  25.0           1083.0   
8         Saul Niguez         3       8  23.0           2925.0   
9           Jan Oblak         1       8  25.0           3330.0   
10       Stefan Savic         2       8  27.0           1398.0   
11             Vitolo         3       8  28.0            753.0   
12       Aritz Aduriz         4      35  37.0            943.0   
13      Yeray Alvarez         2      35  23.0           2613.0   
14   Mikel Balenziaga         2      35  30.0            523.0   
15     Yuri Berchiche         2      35  28.0           3120.0   
16         Ander Capa         2      35  26.0           1642.0   
17      Inigo Cordoba         4      35  21.0           1321.0   
18   Benat Etxebarria         3      35  31.0           2031.0   
19     Cristian Ganea         2      35  26.0             49.0   

    Goles anotados  Asistencias anotadas  Goles anotados esperados  \
0              0.0                   0.0                       0.0   
1              1.0                   1.0                       0.4   
2              3.0                   2.0                       2.2   
3              2.0                   2.0                       1.7   
4             15.0                   8.0                      13.9   
5              3.0                   5.0                       1.5   
6              3.0                   3.0                       3.0   
7              6.0                   1.0                       6.3   
8              4.0                   1.0                       5.0   
9              0.0                   0.0                       0.0   
10             0.0                   2.0                       0.7   
11             0.0                   1.0                       0.9   
12             2.0                   0.0                       3.3   
13             0.0                   0.0                       1.3   
14             0.0                   0.0                       0.0   
15             2.0                   2.0                       1.2   
16             0.0                   0.0                       0.5   
17             0.0                   4.0                       1.3   
18             1.0                   2.0                       0.5   
19             0.0                   0.0                       0.0   

    Goles encajados (Porteros)  Paradas (Porteros)  \
0                          2.0                 5.0   
1                          0.0                 0.0   
2                          0.0                 0.0   
3                          0.0                 0.0   
4                          0.0                 0.0   
5                          0.0                 0.0   
6                          0.0                 0.0   
7                          0.0                 0.0   
8                          0.0                 0.0   
9                         27.0               104.0   
10                         0.0                 0.0   
11                         0.0                 0.0   
12                         0.0                 0.0   
13                         0.0                 0.0   
14                         0.0                 0.0   
15                         0.0                 0.0   
16                         0.0                 0.0   
17                         0.0                 0.0   
18                         0.0                 0.0   
19                         0.0                 0.0   

    Porter

In [49]:
df.to_csv('/Users/alexcar934/Desktop/src/data/processed/Datos.csv')

# El target

### El target será el valor de mercado, estableciendo 4 rangos:
1. De 0 al primer cuartil (C1)
2. De C1 a C2
3. De C2 a C3
4. De C3 a C4
